keystore/v2/keystore/filesystem/keyRingTX.go

```diff
@@ -133,6 +133,34 @@ func (tx *txAddKey) Rollback(ring *KeyRing) error {
 	return nil
 }
 
+type txSetKeys struct {
+	newKeys []asn1.Key
+	current int
+
+	oldKeys    []asn1.Key
+	oldCurrent int
+}
+
+func (tx *txSetKeys) Apply(ring *KeyRing) error {
+	tx.oldKeys = ring.data.Keys
+	tx.oldCurrent = ring.data.Current
+	ring.data.Keys = tx.newKeys
+	ring.data.Current = tx.current
+	return nil
+}
+
+func (tx *txSetKeys) Rollback(ring *KeyRing) error {
+	if ring.data.Current != tx.oldCurrent {
```

Comment:
is not it should be equal to `tx.current`? In `Apply` method we setup `current` for ring. In `Rollback` we should expect any other value from oldCurrent or just setup `oldCurrent`.
p.s. what means "OOO" in `errTxOOORollback`?)

Reply:
> p.s. what means "OOO" in errTxOOORollback?)

cat on keyboard?

Reply:
@Lagovas, I believe you’re right, that line is incorrect. If the current subkey does not change then we’ll get an error here.

> p.s. what means "OOO" in errTxOOORollback?)

Out-of-order.

> cat on keyboard?

I wish it were. The only cat I have is a ceramic one, it can definitely spam one keypress if it falls of the shelf onto the keyboard.

----

I’ve added these checks to verify that we do not roll back changes out-of-order, but this code path is not well tested (if tested at all). So I guess we can drop all these lines and always roll back to whatever values we think are correct ones?

Reply:
then better to rename to full name `errTxOutOfOrderRollback`. Because i'm not sure that anyone of us will be able to repeat what it means after 1-3 months)

and yes, rollback should return state of something to state when it was at start. In this case we store state at start of the transaction and know which state it should take. And it doesn't need to be changed always on rollback. For example with databases, state of table can be unchanged but the transaction rollbacked by some other condition and it's okay. There the same, we don't need check that something changed.